Add gstreamer-gl bindings

ceyusa opened the pull request:
Continuation of issue #126 

This just add the API I could managed to add "as is"

Future iteration will add support for gl crate and native windowing handling.

sdroege commented:
See failures here: https://travis-ci.org/sdroege/gstreamer-rs/jobs/436164556
You'd want to either add `GThread` support to the glib crate (not really difficult or much work), or disable that API for now

ceyusa commented on Gir_GstGL.toml:
```diff
@@ -0,0 +1,225 @@
+    [[object.function]]
+    name = "get_gl_context"
+    # return handle
+    ignore = true
```
I'm thinking in the next iteration for the manual functions.

sdroege replied:
Generally I'm fine with this but these specific functions are needed to do anything meaningful at all

ceyusa commented on Gir_GstGL.toml:
```diff
@@ -0,0 +1,225 @@
+
+    [[object.function]]
+    name = "get_proc_address"
+    ignore = true
```
ok

ceyusa commented on Gir_GstGL.toml:
```diff
@@ -0,0 +1,225 @@
+
+    [[object.function]]
+    name = "thread_add"
+    ignore = true
```
I want to avoid this:

```
+    //pub fn thread_add<P: Into<Option</*Unimplemented*/Fundamental: Pointer>>>(&self, func: /*Unknown conversion*//*Unimplemented*/GLContextThreadFunc, data: P) {                                              
+    //    unsafe { TODO: call ffi::gst_gl_context_thread_add() }
+    //}
```

But, yes, perhaps we would need an unsafe call to offer a callback in rust

sdroege replied:
It could be safe though :)

ceyusa commented on Gir_GstGL.toml:
```diff
@@ -0,0 +1,225 @@
+
+    [[object.function]]
+    name = "error_quark"
+    ignore = true
```
fixed with a manual patch

sdroege replied:
Should be fixed in git master of gst-plugins-base, I cleaned up all this mess

ceyusa commented on Gir_GstGL.toml:
```diff
@@ -0,0 +1,225 @@
+girs_dir = "gir-files"
+library = "GstGL"
+version = "1.0"
+min_cfg_version = "1.0"
```
fixed

ceyusa replied:
but it will need the merge  of https://github.com/sdroege/gstreamer-sys/pull/15

ceyusa commented on Gir_GstGL.toml:
```diff
@@ -0,0 +1,225 @@
+[options]
+girs_dir = "gir-files"
+library = "GstGL"
+version = "1.0"
```
fixed

ceyusa commented on gstreamer-gl/src/auto/gl_color_convert.rs:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    #[cfg(any(feature = "v1_8", feature = "dox"))]
+    pub fn decide_allocation(&self, query: &gst::Query) -> bool {
```
does this mean another manual patch in the autogenerated code?

ceyusa replied:
nope.. :) to fix manually in following iteration

sdroege replied:
Manual code and ignore this via gir in the toml file, yes

ceyusa commented on gstreamer-gl/src/auto/gl_color_convert.rs:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    #[cfg(any(feature = "v1_4", feature = "dox"))]
+    pub fn perform(&self, inbuf: &gst::Buffer) -> Option<gst::Buffer> {
```
would be this a patch in the library gir annotations?

sdroege replied:
Manual code, gir can't express this

ceyusa commented on gstreamer-gl/src/auto/gl_view_convert.rs:
```diff
@@ -0,0 +1,245 @@
+    }
+
+    #[cfg(any(feature = "v1_6", feature = "dox"))]
+    pub fn perform(&self, inbuf: &gst::Buffer) -> Option<gst::Buffer> {
```
I don't get it. ``perform`` can return NULL

sdroege replied:
Can it? OK but does it take ownership of the buffer?

ceyusa replied:
internally it increments the buffer reference, so, iiuc, there is not ownership transference. am i right?

sdroege replied:
Correct, all good then

ceyusa commented on gstreamer-gl/src/auto/gl_view_convert.rs:
```diff
@@ -0,0 +1,245 @@
+    }
+
+    #[cfg(any(feature = "v1_6", feature = "dox"))]
+    pub fn submit_input_buffer(&self, is_discont: bool, input: &gst::Buffer) -> gst::FlowReturn {
```
indeed

sdroege commented on Gir_GstGL.toml:
```diff
@@ -0,0 +1,385 @@
+    "GstGL.GLFormat",
+    "GstGL.GLQueryType",
+    "GstGL.GLSLVersion", # https://github.com/gtk-rs/gir/issues/600
+    "GstGL.GLTextureTarget", # https://github.com/gtk-rs/gir/issues/600
```
You can remove those comments now, it's fixed :)

sdroege commented on Gir_GstGL.toml:
```diff
@@ -0,0 +1,385 @@
+    "GstGL.GLBaseMemoryError", # cannot find function gst_gl_base_buffer_error_quark()
+    "GstGL.GLContextError", # manually add error domain trait
+    "GstGL.GLSLError", # cannot find function gst_glsl_error_quark()
+    "GstGL.GLWindowError", # manually add error domain trait
```
These 4 should be fine with the latest .gir file. Just be careful to *not* include the `get_type()` functions for all these as they will only be available from 1.16.

ceyusa replied:
done! :)

sdroege replied:
Ok then the comments can be removed I guess? :)

sdroege commented:
Manual code should be directly in the src directory, same filename as the autogenerated one. You then include that module in `lib.rs` and it will have the same effect as editing the autogenerated files without overwriting next time

sdroege commented on gstreamer-gl/src/enums.rs:
```diff
@@ -0,0 +1,175 @@
+use GLWindowError;
+
+#[derive(Debug, Eq, PartialEq, Ord, PartialOrd, Hash, Clone, Copy)]
+pub enum GLBaseMemoryError {
```
Why is this still necessary?

ceyusa replied:
because of ``gst_gl_base_memory_error_quark()`` which has to be defined manually because the gir defines ``gst_gl_base_buffer_error_quark()``

sdroege replied:
I fixed that some time last week in GStreamer :)

ceyusa replied:
In master, right? I think we should stick to release 1.14 which needs this hack

sdroege replied:
it's just a change in the gir file, making it point to the right quark functions

ceyusa replied:
with ``xmlstarlet`` in ``fix.sh``?

sdroege replied:
That would work, yes. I fixed it upstream by generating the correct .gir but until we import that new version, fix.sh would be enough I guess

sdroege commented on gstreamer-gl/src/gl_context.rs:
```diff
@@ -0,0 +1,62 @@
+impl GLContext {
+    pub fn new_wrapped(
+        display: &GLDisplay,
+        handle: uintptr_t,
```
This function must be unsafe. If you pass a wrong handle in there, things will crash or worse :)

ceyusa replied:
aha!

Right now I'm cooking an example with this code. I'm aiming to use a  simple window with [glutin](https://crates.io/crates/glutin) and sharing the context to a pipeline of type ``videotestsrc ! glupload ! glcolorconvert ! appsink``. Thus, I'll avoid the guessing :)

sdroege replied:
That example sounds very useful indeed, feel free to put it under the examples subcrate here :)

ceyusa commented:
I have rebased this pull request above https://github.com/sdroege/gstreamer-rs/pull/138
Also I have added a WIP of an example.
The purpose is just for an early review. Perhaps I'm doing something terribly wrong :)

ceyusa commented:
branch updated, over https://github.com/sdroege/gstreamer-rs/pull/138

the examples is still a work in progress and it would need the merge of https://github.com/sdroege/gstreamer-sys/pull/18

sdroege commented on examples/src/bin/glupload.rs:
```diff
@@ -0,0 +1,310 @@
+                };
+
+                let _buffer = if let Some(buffer) = sample.get_buffer() {
+                    buffer
```
The buffer should probably be passed somewhere so that glutin can make use of it?

ceyusa replied:
yep, mapped and paint the texture from it

ceyusa commented on examples/src/bin/glupload.rs:
```diff
@@ -0,0 +1,310 @@
+                if context_type = gst_gl::GL_DISPLAY_CONTEXT_TYPE {
+                }
+                if conteext_type = "gst.gl.app_context" {
+                }
```
yep, create a wrapped display (wip)

JoshuaBatty commented:
Hey @ceyusa & @sdroege. i'm trying to run this pull request on macos and when I run the glupload example i'm getting the following error. 

`panicked at 'Windows can only be created on the main thread on macOS',`

Not sure why this is happening since the example-common.rs should be checking if the os is macos. Just thought id let you know in case you didn't have access to a mac machine for testing. 

BTW, super keen to see this pull request land! I'm not sure what the currently intended use of this crate is without having access to a framebuffer? Is it purely intended to just write your own video players at this stage? Thanks.

sdroege commented:
This is now tracked at https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/merge_requests/131
